### PR TITLE
Add option for using Asynchronous IO in Formatters

### LINKED ISSUE #272: Some problems with Synchronous IO in asp.net core
> In some cases, this error occurs : Microsoft.AspNetCore.Server.Kestrel.Core.BadHttpRequestException: Reading the request body timed out due to data arriving too slowly. See MinRequestBodyDataRate.

some reference:
https://github.com/dotnet/aspnetcore/issues/4707
https://github.com/dotnet/aspnetcore/issues/13834
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AspNetCoreGuidance.md#avoid-using-synchronous-readwrite-overloads-on-httprequestbody-and-httpresponsebody

### LINKED ISSUE #353: Add option for using Asynchronous IO in Formatters
**Is your feature request related to a problem? Please describe.**
See issue #272 

**Describe the solution you'd like**
Add a new property to SparkSettings called: UseAsynchronousIO, if that is true it will pick up the Asynchronous Formatters, otherwise it will use the Synchronous Formatters

**Describe alternatives you've considered**
Removing the Synchronous support in the Formatters altogether. Because we in high-load scenarios have experienced deadlocking this is not an option at the moment

### PR DESCRIPTION
**Pull request by whyfate**
fix #272

### CODE REVIEW

**Comment by kennethmyhra** [comment]
@whyfate seems the XML formatter is broken in this PR. I see from your code that you request a FhirJsonParser in the XML formatter, testing it locally requesting `Accept=application/fhir+xml` returns Json.

You can get the output from the integration tests by clicking the failing integration tests build and download the artfiacts. The annotation artifact gives you a summary of passed and failed tests as well as output from the failing tests.

I have attached the output from the annotation artifact:
```
[
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "PASS: 3228\nFAIL: 12\nERROR: 0\nSKIP: 448",
    "annotation_level": "warning"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "CT01: XML format header mismatch: requested application/fhir+xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "CT02A: XML format param mismatch: requested xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "CT02B: XML format param mismatch: requested text/xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "CT02C: XML format param mismatch: requested application/xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "CT02D: XML format param mismatch: requested XML_FORMAT, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT01: XML format header mismatch: requested application/fhir+xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT02: XML format header mismatch: requested application/fhir+xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT03: Bundle XML format header mismatch: requested application/fhir+xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT04A: Bundle XML format param mismatch: requested xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT04B: Bundle XML format param mismatch: requested text/xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT04C: Bundle XML format param mismatch: requested application/xml, received application/fhir+json",
    "annotation_level": "failure"
  },
  {
    "file": ".github/workflow/integration_tests.yml",
    "line": 1,
    "message": "FT04D: Bundle XML format param mismatch: requested application/fhir+xml, received application/fhir+json",
    "annotation_level": "failure"
  }
]

```

**Comment by whyfate** [comment]
Forgive me for my carelessness.

**Comment by kennethmyhra** [comment]
@whyfate Can you do some performance tests of these changes using https://k6.io/ or something similar? We have had issues with thread-locking previously without buffering and synchronous IO so I want to make sure that bot GET (ResourceJsonOutputFormatter/ResourceXmlOutputFormatter) and PUT/POST (ResourceJsonInputFormatter/ResourceXmlInputFormatter) still can serve under high loads.

I commented regarding this here https://github.com/FirelyTeam/spark/pull/234#issuecomment-650989212. There is also a simple script for GET operations in the thread https://github.com/FirelyTeam/spark/pull/234#issuecomment-651123068.

**Comment by whyfate** [comment]
I tested with https://k6.io/, 
```js
import http from 'k6/http';
import { check } from 'k6';
import { Rate } from 'k6/metrics';

export let errorRate = new Rate("errors");

export let options = {
        insecureSkipTLSVerify: true,
        thresholds: {
                errors: ["rate<0.1"]    // 10% errors
        },
        stages: [
                { duration: '1m', target: 50 },
                { duration: '1m', target: 50 }
        ]
};

export default function() {
        var url = 'http://localhost:5000/fhir/Organization/000015'
        var params = {
                headers: {
                        'Accept': 'application/xml'
                }
        };
        let res = http.get(url, params);
        const result = check(res, {
                "is status 200": r => r.status == 200
        });
        errorRate.add(!result);
}
```
and the results：
```
default ✓ [ 100% ] 00/50 VUs  2m0s

     ✓ is status 200

     checks.....................: 100.00% ✓ 67361 ✗ 0    
     data_received..............: 91 MB   757 kB/s
     data_sent..................: 9.2 MB  76 kB/s
   ✓ errors.....................: 0.00%   ✓ 0     ✗ 67361
     http_req_blocked...........: avg=6.93µs   min=3.4µs  med=5.5µs   max=2.88ms   p(90)=6.6µs    p(95)=7.7µs   
     http_req_connecting........: avg=882ns    min=0s     med=0s      max=2.77ms   p(90)=0s       p(95)=0s      
     http_req_duration..........: avg=66.56ms  min=6.64ms med=59.16ms max=411.59ms p(90)=128.18ms p(95)=152.12ms
     http_req_receiving.........: avg=114.85µs min=40.7µs med=89.5µs  max=52.12ms  p(90)=159.3µs  p(95)=177.9µs 
     http_req_sending...........: avg=35.46µs  min=18.3µs med=28µs    max=2.97ms   p(90)=61.2µs   p(95)=71.6µs  
     http_req_tls_handshaking...: avg=0s       min=0s     med=0s      max=0s       p(90)=0s       p(95)=0s      
     http_req_waiting...........: avg=66.41ms  min=6.46ms med=59.01ms max=411.42ms p(90)=128.03ms p(95)=151.93ms
     http_reqs..................: 67361   560.563648/s
     iteration_duration.........: avg=66.81ms  min=6.86ms med=59.41ms max=411.83ms p(90)=128.42ms p(95)=152.36ms
     iterations.................: 67361   560.563648/s
     vus........................: 50      min=1   max=50 
     vus_max....................: 50      min=50  max=50 
```
docker stats
```
911f50a37d34        spark               147.42%             281.3MiB / 7.781GiB   3.53%               234MB / 267MB       100MB / 0B          23
d96a8ac46e6c        mongodb        225.21%             128.8MiB / 7.781GiB   1.62%               43.7MB / 203MB      88.9MB / 664MB      84
```

**Comment by brianpos** [comment]
p.s. I'm working on a new XML Serializer/Parser that can use the Async features of the XmlReader/Writier (though only in R4)
It will be going into my facade layer that you're welcome to use - and likely a NuGet package by itself.
Still deciding if I should drain the incoming stream to a memory stream (async) and then use the sync version of my parser, or just use the async parser.

**Comment by brianpos** [comment]
The code for my formatter using the new XML Reader/Writer is in this branch if you wanted to try it out locally
https://github.com/brianpos/fhir-net-web-api/blob/feature/r4-async-serializers/src/Hl7.Fhir.WebApi.AspNetCore/XmlFhirFormatter.cs
(I'm still testing that code, but appears to be working pretty well)
Would be interested to see your stats from above with Spark using the custom serializer, and something long the lines of this code in my XmlFormatter class (which was originally from Spark in the pre DSTU2 days)

**Comment by whyfate** [comment]
> The code for my formatter using the new XML Reader/Writer is in this branch if you wanted to try it out locally
> https://github.com/brianpos/fhir-net-web-api/blob/feature/r4-async-serializers/src/Hl7.Fhir.WebApi.AspNetCore/XmlFhirFormatter.cs
> (I'm still testing that code, but appears to be working pretty well)
> Would be interested to see your stats from above with Spark using the custom serializer, and something long the lines of this code in my XmlFormatter class (which was originally from Spark in the pre DSTU2 days)

Cool, expect it to merge into fhirly-net-sdk.

**Comment by brianpos** [comment]
Using your K6 load test locally with my demofilesystemserver in that project above (3.8k test resource from local disk):
```
running (2m00.0s), 00/50 VUs, 451737 complete and 0 interrupted iterations
default ✓ [======================================] 00/50 VUs  2m0s.

     ✓ is status 200

     checks.........................: 100.00% ✓ 451737 ✗ 0
     data_received..................: 1.9 GB  16 MB/s
     data_sent......................: 60 MB   497 kB/s
   ✓ errors.........................: 0.00%   ✓ 0      ✗ 451737
     http_req_blocked...............: avg=4.8µs    min=0s med=0s     max=25ms     p(90)=0s      p(95)=0s
     http_req_connecting............: avg=156ns    min=0s med=0s     max=4ms      p(90)=0s      p(95)=0s
     http_req_duration..............: avg=9.72ms   min=0s med=8.01ms max=283.4ms  p(90)=17.99ms p(95)=22.01ms
       { expected_response:true }...: avg=9.72ms   min=0s med=8.01ms max=283.4ms  p(90)=17.99ms p(95)=22.01ms
     http_req_failed................: 0.00%   ✓ 0      ✗ 451737
     http_req_receiving.............: avg=171.96µs min=0s med=0s     max=42ms     p(90)=997.4µs p(95)=1ms
     http_req_sending...............: avg=24.78µs  min=0s med=0s     max=29ms     p(90)=0s      p(95)=0s
     http_req_tls_handshaking.......: avg=0s       min=0s med=0s     max=0s       p(90)=0s      p(95)=0s
     http_req_waiting...............: avg=9.52ms   min=0s med=8ms    max=283.4ms  p(90)=17.99ms p(95)=22ms
     http_reqs......................: 451737  3763.694479/s
     iteration_duration.............: avg=9.95ms   min=0s med=8.99ms max=285.39ms p(90)=18ms    p(95)=22.99ms
     iterations.....................: 451737  3763.694479/s
     vus............................: 50      min=1    max=50
     vus_max........................: 50      min=50   max=50
```
With compression on it slows things up a little, but that would depend on where your bottleneck is ;)

**Comment by brianpos** [comment]
> > The code for my formatter using the new XML Reader/Writer is in this branch if you wanted to try it out locally
> > https://github.com/brianpos/fhir-net-web-api/blob/feature/r4-async-serializers/src/Hl7.Fhir.WebApi.AspNetCore/XmlFhirFormatter.cs
> > (I'm still testing that code, but appears to be working pretty well)
> > Would be interested to see your stats from above with Spark using the custom serializer, and something long the lines of this code in my XmlFormatter class (which was originally from Spark in the pre DSTU2 days)
> 
> Cool, expect it to merge into fhirly-net-sdk.

I think it will remain an external option, as it uses a different approach.
But I would love to have a way to switch them, and Gino from MS is working on a json version similar to mine.

**Comment by kennethmyhra** [comment]
@brianpos if there were some common interface between the firely-net-sdk and Gino's + your parser it would be possible to switch them using dependency injection

@whyfate as for your original request I think we will need to look at some way of controlling through the Settings interface whether you would like to use the buffering approach or the approach you are outlining in this PR.

**Comment by whyfate** [comment]
> @brianpos if there were some common interface between the firely-net-sdk and Gino's + your parser it would be possible to switch them using dependency injection
> 
> @whyfate as for your original request I think we will need to look at some way of controlling through the Settings interface whether you would like to use the buffering approach or the approach you are outlining in this PR.

My original idea was to use asynchronous io as much as possible instead of synchronization io, because `Spark` used `IAsyncFhirService`, and I hope to use the official method provided in the firely-net-sdk to implement it.

**Comment by kennethmyhra** [comment]
@whyfate I can pull this, but I need us to keep the Synchronous IO path while making it possible through configuration to use the Asynchronous IO path

Adding a new property to SparkSettings called: UseAsynchronousIO, if that is true it will use the new path outlined by you here and if it is false use the old path

Is that something you could put together?

**Comment by whyfate** [comment]
@kennethmyhra I get it.

File: src/Spark.Engine/Formatters/NetCore/AsyncResourceJsonInputFormatter.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            _parser = new FhirJsonParser();
+
+            SupportedEncodings.Clear();
+            SupportedEncodings.Add(Encoding.UTF8);
+
+            SupportedMediaTypes.Add("application/json");
+            SupportedMediaTypes.Add("application/fhir+json");
+            SupportedMediaTypes.Add("application/json+fhir");
+            SupportedMediaTypes.Add("text/json");
+        }
```
Comment (kennethmyhra) [inline src/Spark.Engine/Formatters/NetCore/AsyncResourceJsonInputFormatter.cs]: Can you remove this. Should not be needed for the new Formatters

File: src/Spark.Engine/Formatters/NetCore/AsyncResourceXmlInputFormatter.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+            SupportedEncodings.Clear();
+            SupportedEncodings.Add(Encoding.UTF8);
+
+            SupportedMediaTypes.Add("application/xml");
+            SupportedMediaTypes.Add("application/fhir+xml");
+            SupportedMediaTypes.Add("application/xml+fhir");
+            SupportedMediaTypes.Add("text/xml");
+            SupportedMediaTypes.Add("text/xml+fhir");
+        }
```
Comment (kennethmyhra) [inline src/Spark.Engine/Formatters/NetCore/AsyncResourceXmlInputFormatter.cs]: Can you remove this. Should not be needed for the new Formatters

**Review by kennethmyhra — CHANGES_REQUESTED** [review]
Thanks, looks good! Just remove the two constructors I marked in the new formatters and we are good to go

**Comment by whyfate** [comment]
So careless!!!

**Comment by kennethmyhra** [comment]
Haha, no! It's an easy mistake to make

**Comment by whyfate** [comment]
Thanks

**Review by kennethmyhra — APPROVED** [review]

**Comment by kennethmyhra** [comment]
Thanks @whyfate!

**Pull request merged**